tests/zfs-tests/tests/functional/dedup/dedup_quota.ksh

```diff
@@ -0,0 +1,221 @@
+POOL="dedup_pool"
+
+ORIG_TXG_TIMEOUT=$(get_tunable TXG_TIMEOUT)
+
+log_note ddt_zap_default_bs = $ORIG_DDT_ZAP_DEFAULT_BS
+
+function cleanup
+{
+	if poolexists $POOL ; then
+		destroy_pool $POOL
+	fi
+	log_must rm -fd $VDEV_GENERAL $VDEV_DEDUP $MOUNTDIR
+	log_must set_tunable32 TXG_TIMEOUT $ORIG_TXG_TIMEOUT
+}
+
+
+function do_clean
+{
+	log_must zpool destroy $POOL
+	log_must rm -fd $VDEV_GENERAL $VDEV_DEDUP $MOUNTDIR
+}
+
+function do_setup
+{
+	# Create a general VDEV and special dedup VDEV
+	log_must truncate -s 5G $VDEV_GENERAL
+	log_must truncate -s 200M $VDEV_DEDUP
+	# Use 'xattr=sa' to prevent selinux xattrs influencing our accounting
+	log_must zpool create -o ashift=12 -f -O xattr=sa -m $MOUNTDIR $POOL $VDEV_GENERAL
+	log_must zfs set dedup=on $POOL
+	log_must set_tunable32 TXG_TIMEOUT 600
+}
+
+function dedup_table_size
+{
+	zpool get -Hp -o value dedup_table_size $POOL
+}
+
+function dedup_table_quota
+{
+	zpool get -Hp -o value dedup_table_quota $POOL
+}
+
+function ddt_entries
+{
+	entries=$(zpool status -D $POOL | \
+		grep "dedup: DDT entries" | awk '{print $4}')
+
+	echo ${entries}
+}
+
+function ddt_add_entry
+{
+	count=$1
+	offset=$2
+	expand=$3
+
+	if [ -z "$offset" ]; then
+		offset=1
+	fi
+
+	for i in {$offset..$count}; do
+		echo "$i" > $MOUNTDIR/dedup-$i.txt
+	done
+	zpool sync $POOL
+
+	log_note range $offset - $(( count + offset - 1 ))
+	log_note ddt_add_entry got $(ddt_entries)
+}
+
+# Create 6000 entries over three syncs
+function ddt_nolimit
+{
+	do_setup
+
+	log_note base ddt entries is $(ddt_entries)
+
+	ddt_add_entry 1
+	ddt_add_entry 100
+	ddt_add_entry 101 5000
+	ddt_add_entry 5001 6000
+
+	log_must test $(ddt_entries) -eq 6000
+
+	do_clean
+}
+
+function ddt_limit
+{
+	do_setup
+
+	log_note base ddt entries is $(ddt_entries)
+
+	log_must zpool set dedup_table_quota=32768 $POOL
+	ddt_add_entry 100
+
+	# it's possible to exceed dedup_table_quota over a single transaction,
+	# ensure that the threshold has been exceeded
+	cursize=$(dedup_table_size)
+	log_must test $cursize -gt $(dedup_table_quota)
+
+	# count the entries we have
+	log_must test $(ddt_entries) -ge 100
+
+	# attempt to add new entries
+	ddt_add_entry 101 200
+	log_must test $(ddt_entries) -eq 100
+	log_must test $cursize -eq $(dedup_table_size)
+
+	# remove all but one entry
+	for i in {2..100}; do
+		rm $MOUNTDIR/dedup-$i.txt
+	done
+	zpool sync
+
+	log_must test $(ddt_entries) -eq 1
+	log_must test $cursize -gt $(dedup_table_size)
+	cursize=$(dedup_table_size)
+
+	log_must zpool set dedup_table_quota=none $POOL
+
+	# create more entries
+	zpool status -D $POOL
+	ddt_add_entry 101 200
+	zpool sync
+
+	log_must test $(ddt_entries) -eq 101
+	log_must test $cursize -lt $(dedup_table_size)
+
+	do_clean
+}
+
+function ddt_dedup_vdev_limit
+{
+	do_setup
+
+	# add a dedicated dedup VDEV and enable an automatic quota
+	log_must zpool add $POOL dedup $VDEV_DEDUP
+	log_must zpool set dedup_table_quota=auto $POOL
```

Comment:
We should also test the `zpool add $POOL special $VDEV_DEDUP` behavior for set values and auto.

Reply:
made the vdev class choice, `dedup` or `special`, a random choice so we get coverage over time for both.